File: language/compiler/src/com/kubadziworski/parsing/visitor/expression/ExpressionVisitor.java
package com.kubadziworski.parsing.visitor.expression;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser.AddContext;
import jiva.parser.JivaParser.ConditionalExpressionContext;
import jiva.parser.JivaParser.ConstructorCallContext;
import jiva.parser.JivaParser.DivideContext;
import jiva.parser.JivaParser.FunctionCallContext;
import jiva.parser.JivaParser.MultiplyContext;
import jiva.parser.JivaParser.SubstractContext;
import jiva.parser.JivaParser.SupercallContext;
import jiva.parser.JivaParser.ValueContext;
import jiva.parser.JivaParser.VarReferenceContext;
import com.kubadziworski.domain.node.expression.*;
import com.kubadziworski.domain.scope.Scope;
import com.kubadziworski.parsing.visitor.expression.function.CallExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

/**
 * Created by kuba on 02.04.16.
 */
public class ExpressionVisitor extends JivaBaseVisitor<Expression> {

    private final ArithmeticExpressionVisitor arithmeticExpressionVisitor;
    private final VariableReferenceExpressionVisitor variableReferenceExpressionVisitor;
    private final ValueExpressionVisitor valueExpressionVisitor;
    private final CallExpressionVisitor callExpressionVisitor;
    private final ConditionalExpressionVisitor conditionalExpressionVisitor;

    public ExpressionVisitor(Scope scope) {
        arithmeticExpressionVisitor = new ArithmeticExpressionVisitor(this);
        variableReferenceExpressionVisitor = new VariableReferenceExpressionVisitor(scope);
        valueExpressionVisitor = new ValueExpressionVisitor();
        callExpressionVisitor = new CallExpressionVisitor(this, scope);
        conditionalExpressionVisitor = new ConditionalExpressionVisitor(this);
    }

    @Override
    public Expression visitVarReference(@NotNull VarReferenceContext ctx) {
        return variableReferenceExpressionVisitor.visitVarReference(ctx);
    }

    @Override
    public Expression visitValue(@NotNull ValueContext ctx) {
        return valueExpressionVisitor.visitValue(ctx);
    }

    @Override
    public Expression visitFunctionCall(@NotNull FunctionCallContext ctx) {
        return callExpressionVisitor.visitFunctionCall(ctx);
    }

    @Override
    public Expression visitConstructorCall(@NotNull ConstructorCallContext ctx) {
        return callExpressionVisitor.visitConstructorCall(ctx);
    }

    @Override
    public Expression visitSupercall(@NotNull SupercallContext ctx) {
        return callExpressionVisitor.visitSupercall(ctx);
    }

    @Override
    public Expression visitAdd(@NotNull AddContext ctx) {

        return arithmeticExpressionVisitor.visitAdd(ctx);
    }

    @Override
    public Expression visitMultiply(@NotNull MultiplyContext ctx) {

        return arithmeticExpressionVisitor.visitMultiply(ctx);
    }

    @Override
    public Expression visitSubstract(@NotNull SubstractContext ctx) {

        return arithmeticExpressionVisitor.visitSubstract(ctx);
    }

    @Override
    public Expression visitDivide(@NotNull DivideContext ctx) {

        return arithmeticExpressionVisitor.visitDivide(ctx);
    }

    @Override
    public ConditionalExpression visitConditionalExpression(@NotNull ConditionalExpressionContext ctx) {
        return conditionalExpressionVisitor.visitConditionalExpression(ctx);
    }
}

File: language/compiler/src/parsing/visitor/expression/ExpressionVisitor.kt
package jiva.parsing.visitor.expression

import jiva.domain.node.expression.ConditionalExpression
import jiva.domain.node.expression.Expression
import jiva.domain.scope.Scope
import jiva.parsing.visitor.expression.function.CallExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.*

/**
 * Created by kuba on 02.04.16.
 */
class ExpressionVisitor(scope: Scope) : JivaBaseVisitor<Expression?>() {

    private val arithmeticExpressionVisitor: ArithmeticExpressionVisitor = ArithmeticExpressionVisitor(this)
    private val variableReferenceExpressionVisitor: VariableReferenceExpressionVisitor = VariableReferenceExpressionVisitor(scope)
    private val valueExpressionVisitor: ValueExpressionVisitor = ValueExpressionVisitor()
    private val callExpressionVisitor: CallExpressionVisitor = CallExpressionVisitor(this, scope)
    private val conditionalExpressionVisitor: ConditionalExpressionVisitor = ConditionalExpressionVisitor(this)

    override fun visitVarReference(ctx: VarReferenceContext): Expression {
        return variableReferenceExpressionVisitor.visitVarReference(ctx)
    }

    override fun visitValue(ctx: ValueContext): Expression {
        return valueExpressionVisitor.visitValue(ctx)
    }

    override fun visitFunctionCall(ctx: FunctionCallContext): Expression {
        return callExpressionVisitor.visitFunctionCall(ctx)
    }

    override fun visitConstructorCall(ctx: ConstructorCallContext): Expression {
        return callExpressionVisitor.visitConstructorCall(ctx)
    }

    override fun visitSupercall(ctx: SupercallContext): Expression {
        return callExpressionVisitor.visitSupercall(ctx)
    }

    override fun visitAdd(ctx: AddContext): Expression {
        return arithmeticExpressionVisitor.visitAdd(ctx)
    }

    override fun visitMultiply(ctx: MultiplyContext): Expression {
        return arithmeticExpressionVisitor.visitMultiply(ctx)
    }

    override fun visitSubstract(ctx: SubstractContext): Expression {
        return arithmeticExpressionVisitor.visitSubstract(ctx)
    }

    override fun visitDivide(ctx: DivideContext): Expression {
        return arithmeticExpressionVisitor.visitDivide(ctx)
    }

    override fun visitConditionalExpression(ctx: ConditionalExpressionContext): ConditionalExpression {
        return conditionalExpressionVisitor.visitConditionalExpression(ctx)
    }

}